### ISSUE TITLE
Lost of constraint with conditionals

### ISSUE THREAD

**Issue by cbchouinard** [gss/engine#167]
I have a strange behaviour with conditionals when resizing the window. I loose the `width` constraint on `.content`. 

On the first load of the page it works fine, if I reduce the size of the window to hit the `@else` it works fine too, but when I augment the size of the window and get into the `@if` conditional then `.content[width]` is not constrain anymore and the div expand according to normal flow. 

```
...

@if ::window[width] >= $breakpoint1 {

      header > h1 {
        @h |-(&)-| in(^) gap($col-size);
      }

      .content[width] <= $max-p-width - $baseline * 4 !strong;
      
      device-out-gap >= 0 !require;
      device-in-gap == 50 !strong;

      @v |-(header)-(.content)-| in(&) gap($baseline) outer-gap($v-outer-gap) !strong;
      @h |-(#spaceman)-(.content)-| in(&) gap(device-in-gap) outer-gap(device-out-gap) !strong;


      #spaceman[opacity] == 0.5;

    } @else {

      #spaceman[opacity] == 0.15;
      .content[center-x] == ^center-x !strong;
      #spaceman[center-x] == .content[center-x];

      header > h1, .content {
        @h |-(&)-| in(^) gap($col-size / 2);
      }

      .content[width] <= $max-p-width - $baseline;

      @v |-(header)-(.content)-| in(&) gap($baseline / 2) outer-gap($baseline);

     
    }

...
```

I'll try to do a codepen later to reproduce the problem.

**Comment by Inviz** [comment]
There're two possible problems, and it's important to understand which is the case for you:

1) Constraint set is different due to some bug in gss document code. Check numbers in graph panel when you've got logging turned on. It should have the same number when switching to another breakpoint and going back.

2) Constraint set is the same, but solution is different. This can happen when using inequalities. GSS attempts to drop cassowary state when switching branches, but in some complex situations it still would not produce a desirable solution. Sometimes it happens due to use of inequalities, which don't provide a strong enough effect when applied iteratively.

In either case we need to reduce the problem to the smallest possible setup that exhibits the problem.

**Comment by primedirective** [comment]
I think I'm running into something like this. So I have a viewport for a grid of items, like...

    "grid-viewport" {
      ::[top] == ($ #toolbar)[bottom] !require;
      ::[right] == ::window[right] !strong;
      ::[left] == "side-menu"[right] !strong;

      ::[width] == ::window[width] - "side-menu"[width] !require;
      ::[height] == ::window[height] - ($ #toolbar)[height] !require;
    }

And there's a whole bunch of items I'm laying out in a grid, similar to the way it's done in the documentation. Here's the whole splat of items...

    "area1...24" {
      width: == "grid-viewport"[width] / 4; /* Number of Columns */
      height: == "grid-viewport"[height] / 2.5; /* Show 2.5 rows in viewport */
    }

And then I just lay them out all the way to 24 and the page is scrollable.

What I want to do is have the height on "area1...24" be dependent on ::window[width] (for example changing it to divide the viewport height by 4 instead of 2.5 if the window width is <= 980). That seems to work if I load the page up on any particular size, but if I *resize* the page to meet the other condition, it seems to go crazy and we completely lose height.

What doesn't seem to work is adding a conditional on ::window[width] either inside the virtual splat, or trying to do it outside. If I load the page with a small width, it will seem to work, until I make it larger than 980. Even if I literally declare what I want to happen when it's greater than 980, it still doesn't seem to work on page resize.

Any thoughts on this?

**Comment by Inviz** [comment]
Hello. Thanks for detailed question. Unfortunately it's still not clear to me what you would like to achieve with 24 virtuals. Do you bind their positions? Because the way you use it in example just iterates them, and does not imply position binding. 

In our internal projects we try to avoid using virtuals for everything that has dedicated elements. For cases where we need a simple 10-20 column grid we use simple math.

**Comment by primedirective** [comment]
Hi @Inviz. Yep, they are bound. Kind of messy, but here's what I've got:

    /* - 3 ROWS, 4 COLS - */

    "area1...24" {
      width: == "grid-viewport"[width] / 4;
      height: == "grid-viewport"[height] / 2.5;
    } /* set up the grid subdivisions based on grid viewport width */
      /* this must reflect total number of items */

    "area1...4" {
      top: == "grid-viewport"[top] !strong;
      left: == &:previous[right];
    } /* top boundary, left neighbor */

    "area5...8" {
      left: == &:previous[right];
    } /* left neighbor */

    "area9...12" {
      left: == &:previous[right];
    } /* left neighbor */

    "area1", "area5", "area9", "area13", "area17", "area21" {
      left: == "grid-viewport"[left];
      top: == &:previous[bottom];
    } /* left boundary, first column binding */
      /* ONE */

    "area4", "area8", "area12", "area16", "area20", "area24" {
      right: == "grid-viewport"[right];
      top: == &:previous[bottom];
    } /* right boundary, fourth column binding */
      /* FOUR */

    "area2", "area6", "area10", "area14", "area18", "area22" {
      top: == &:previous[bottom];
    } /* second column binding */
      /* TWO */

    "area3", "area7", "area11", "area15", "area19", "area23" {
      top: == &:previous[bottom];
    } /* third column binding */
      /* THREE */

    "area13...16" {
      left: == &:previous[right];
    } /* 1st below the fold column,  left neighbors */

    "area17...20" {
      left: == &:previous[right];
    } /* 2nd below the fold column,  left neighbors */

    "area21...24" {
      left: == &:previous[right];
    } /* 2nd below the fold column,  left neighbors */

    /* -------------------- */

Basically, the way things are laid out, you get 4 columns and a bunch of rows. The idea is that as the page is resized, the virtuals change size proportionally. What I want to do is have it change the factor it's dividing the area1...24 height by. Sometimes that's happening, but most of the time, the height just gets completely lost. Not sure what's going on there...

**Comment by paulyoung** [comment]
@primedirective you may prefer to write things like this:

```
@h |("area1...4")| in(::window);
@h |("area5...8")| in(::window);
@h |("area9...12")| in(::window);
@h |("area13...16")| in(::window);
@h |("area17...20")| in(::window);
@h |("area21...24")| in(::window);

@v |("area1")("area5")("area9")("area13")("area17")("area21") in(::window);
@v |("area2")("area6")("area10")("area14")("area18")("area22") in(::window);
@v |("area3")("area7")("area11")("area15")("area19")("area23") in(::window);
@v |("area4")("area8")("area12")("area16")("area20")("area24") in(::window);
```

**Comment by primedirective** [comment]
@paulyoung Tried this out and it ultimately ended up producing the same behavior. So I ended up doing essentially this, but pinning it to the actual tile divs instead of the virtuals and it ended up working out.

Thanks for your help!
